examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/config/SPIBasedShardingRawJavaConfiguration.java

```diff
@@ -0,0 +1,73 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.example.extension.spibased.sharding.raw.jdbc.config;
+
+import org.apache.shardingsphere.driver.api.ShardingSphereDataSourceFactory;
+import org.apache.shardingsphere.example.config.ExampleConfiguration;
+import org.apache.shardingsphere.example.core.api.DataSourceUtil;
+import org.apache.shardingsphere.infra.config.algorithm.AlgorithmConfiguration;
+import org.apache.shardingsphere.sharding.api.config.ShardingRuleConfiguration;
+import org.apache.shardingsphere.sharding.api.config.rule.ShardingTableRuleConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.keygen.KeyGenerateStrategyConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.sharding.StandardShardingStrategyConfiguration;
+
+import javax.sql.DataSource;
+import java.sql.SQLException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+public final class SPIBasedShardingRawJavaConfiguration implements ExampleConfiguration {
+    
+    @Override
+    public DataSource getDataSource() throws SQLException {
+        return ShardingSphereDataSourceFactory.createDataSource(createDataSourceMap(), Collections.singletonList(createShardingRuleConfiguration()), new Properties());
+    }
+    
+    private Map<String, DataSource> createDataSourceMap() {
+        Map<String, DataSource> result = new HashMap<>();
+        result.put("demo_ds_0", DataSourceUtil.createDataSource("demo_ds_0"));
+        result.put("demo_ds_1", DataSourceUtil.createDataSource("demo_ds_1"));
+        return result;
+    }
+    
+    private ShardingRuleConfiguration createShardingRuleConfiguration() {
+        ShardingRuleConfiguration result = new ShardingRuleConfiguration();
+        result.getTables().add(getOrderTableRuleConfiguration());
+        result.getTables().add(getOrderItemTableRuleConfiguration());
+        result.getBindingTableGroups().add("t_order, t_order_item"); 
+        result.getBroadcastTables().add("t_address");
+        result.setDefaultDatabaseShardingStrategy(new StandardShardingStrategyConfiguration("user_id", "spi_based"));
+        result.getShardingAlgorithms().put("spi_based", new AlgorithmConfiguration("SPI_BASED", new Properties()));
+        result.getKeyGenerators().put("snowflake", new AlgorithmConfiguration("SNOWFLAKE", new Properties()));
+        return result;
+    }
+    
+    private ShardingTableRuleConfiguration getOrderTableRuleConfiguration() {
+        ShardingTableRuleConfiguration result = new ShardingTableRuleConfiguration("t_order", "demo_ds_${0..1}.t_order_${[0, 1]}");
```

Comment:
Hi @lushaorong,
Nice to see you again, here's a little problem:
The sharding type you specified is `SHARDING_DATABASES`, but the actualDataNodes here contains the logic of sharding tables, which is obviously incorrect.

It is recommended that you test it before submitting a PR. If you have any questions, welcome to communicate.

Reply:
SHARDING_DATABASES_AND_TABLES is more suitable? 
But I refer to [here](https://github.com/apache/shardingsphere/blob/master/examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/SPIBasedShardingRawYamlConfigurationExample.java), it contains the logic of sharding tables and use SHARDING_DATABASES.Is it right？

Reply:
Either way, the configuration should match the name.
But you used `SHARDING_DATABASES`, and dataNodes `"demo_ds_${0..1}.t_order_${[0, 1]}"` is for `SHARDING_DATABASES_AND_TABLES`.

This is not a match.

Reply:
You're right.
I'll change it to SHARDING_DATABASES_AND_TABLES

Reply:
It is still SHARDING_DATABASES and I change the rule `demo_ds_${0..1}.t_order_item_${[0, 1]}` to `demo_ds_${0..1}.t_order_item`